src/System.Private.CoreLib/shared/System/Number.Parsing.cs

```diff
@@ -961,13 +961,13 @@ private static bool NumberBufferToDouble(ref NumberBuffer number, ref double val
         private static class DoubleHelper
         {
             public static unsafe uint Exponent(double d) =>
-                (*((uint*)&d + 1) >> 20) & 0x000007ff;
+                (*((uint*)&d + (BitConverter.IsLittleEndian ? 1 : 0)) >> 20) & 0x000007ff;
 
             public static unsafe ulong Mantissa(double d) =>
                 *((ulong*)&d) & 0x000fffffffffffff;
 
             public static unsafe bool Sign(double d) =>
-                (*((uint*)&d + 1) >> 31) != 0;
+                (*((uint*)&d + (BitConverter.IsLittleEndian ? 1 : 0)) >> 31) != 0;
```

Comment:
Does the JIT successfully see IsLittleEndian as a constant and elide the branch such that the asm ends up being identical to what it is today on little-endian machines?

Reply:
Yup: #9789

Reply:
I see something strange here.

Old version:
```asm
G_M7422_IG01:
       50                   push     rax
       C5F877               vzeroupper 

G_M7422_IG02:
       C4E17B110424         vmovsd   qword ptr [rsp], xmm0
       8B442404             mov      eax, dword ptr [rsp+04H]
       C1E81F               shr      eax, 31
       85C0                 test     eax, eax
       0F95C0               setne    al
       0FB6C0               movzx    rax, al

G_M7422_IG03:
       4883C408             add      rsp, 8
       C3                   ret      
```

New version:
```asm

G_M7422_IG01:
       50                   push     rax
       C5F877               vzeroupper 

G_M7422_IG02:
       E827E3FFFF           call     Program:Sign2(double):bool
       90                   nop      

G_M7422_IG03:
       4883C408             add      rsp, 8
       C3                   ret      
```
`[FAILED: unprofitable inline]`

<details>
  <summary>c# Code</summary>

```c#
using System;
using System.Runtime.CompilerServices;

namespace ConsoleApp1
{
    class Program
    {
        static int Main(string[] args)
        {
            Console.WriteLine(BitConverter.IsLittleEndian);
            double d = 42d;

            bool b1 = A(d);
            bool b2 = B(d);

            return b1 == b2 ? 0 : 1;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        private static bool A(double d)
        {
            return Sign1(d);
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        private static bool B(double d)
        {
            return Sign2(d);
        }

        private static unsafe bool Sign1(double d) => (*((uint*)&d + 1) >> 31) != 0;

        private static unsafe bool Sign2(double d) => (*((uint*)&d + (BitConverter.IsLittleEndian ? 1 : 0)) >> 31) != 0;
    }
}
```
</details>

@mikedn did you mean this?

Reply:
@gfoidl Something like that. I didn't look at the impact this has on inlining but it does mess with the address calculation:
Original:
```asm
       8B44240C             mov      eax, dword ptr [rsp+0CH]
       C1E81F               shr      eax, 31
       85C0                 test     eax, eax
       0F95C0               setne    al
       0FB6C0               movzx    rax, al
```
New:
```asm
       488D442408           lea      rax, bword ptr [rsp+08H]
       BA01000000           mov      edx, 1
       4863D2               movsxd   rdx, edx
       8B0490               mov      eax, dword ptr [rax+4*rdx]
       C1E81F               shr      eax, 31
       85C0                 test     eax, eax
       0F95C0               setne    al
       0FB6C0               movzx    rax, al
```
If may be better to use
```C#
return IsLittleEndian ? ((*((uint*)&d + 1) >> 31) != 0) : ((*((uint*)&d) >> 31) != 0);
```
Though that may still impact inlining.

Reply:
> If may be better to use

Right, same asm as old one, but inlining impact. `AggressiveInlining` would do it...